mars/dataframe/expressions/arithmetic/core.py

```diff
@@ -0,0 +1,358 @@
+        return self._column_min_close
+
+    @property
+    def column_max(self):
+        return self._column_max
+
+    @property
+    def column_max_close(self):
+        return self._column_max_close
+
+    @property
+    def column_shuffle_size(self):
+        return self._column_shuffle_size
+
+
+class DataFrameIndexAlignReduce(ShuffleReduce, DataFrameOperandMixin):
+    _op_type_ = OperandDef.DATAFRAME_INDEX_ALIGN_REDUCE
+
+    _input = KeyField('input')
+
+    def __init__(self, shuffle_key=None, sparse=None, **kw):
+        super(DataFrameIndexAlignReduce, self).__init__(
+            _shuffle_key=shuffle_key, _sparse=sparse, **kw)
+
+    def _set_inputs(self, inputs):
+        super(DataFrameIndexAlignReduce, self)._set_inputs(inputs)
+        self._input = self._inputs[0]
+
+    def calc_shape(self, *inputs_shape):
+        return self.outputs[0].shape
+
+
+class DataFrameBinOpMixin(DataFrameOperandMixin):
+    __slots__ = ()
+
+    @classmethod
+    def _check_overlap(cls, chunk_index_min_max):
+        for j in range(len(chunk_index_min_max) - 1):
+            # overlap only if the prev max is close and curr min is close
+            # and they are identical
+            prev_max, prev_max_close = chunk_index_min_max[j][2:]
+            curr_min, curr_min_close = chunk_index_min_max[j + 1][:2]
+            if prev_max_close and curr_min_close and prev_max == curr_min:
+                return True
+        return False
+
+    @classmethod
+    def _get_chunk_index_min_max(cls, df, index_type, axis):
+        index = getattr(df, index_type)
+        if not index.is_monotonic_increasing_or_decreasing:
+            return
+
+        chunk_index_min_max = []
+        for i in range(df.chunk_shape[axis]):
+            chunk_idx = [0, 0]
+            chunk_idx[axis] = i
+            chunk = df.cix[tuple(chunk_idx)]
+            chunk_index = getattr(chunk, index_type)
+            min_val = chunk_index.min_val
+            min_val_close = chunk_index.min_val_close
+            max_val = chunk_index.max_val
+            max_val_close = chunk_index.max_val_close
+            if min_val is None or max_val is None:
+                return
+            chunk_index_min_max.append((min_val, min_val_close, max_val, max_val_close))
+
+        if index.is_monotonic_decreasing:
+            return list(reversed(chunk_index_min_max)), False
+
+        if cls._check_overlap(chunk_index_min_max):
+            return
+        return chunk_index_min_max, True
+
+    @classmethod
+    def _get_dtypes(cls, dtypes, column_min_max):
+        l_filter = operator.ge if column_min_max[1] else operator.gt
+        l = l_filter(dtypes.index, column_min_max[0])
+        r_filter = operator.le if column_min_max[3] else operator.lt
+        r = r_filter(dtypes.index, column_min_max[2])
+        f = l & r
+        return dtypes[f]
+
+    @classmethod
+    def _gen_out_chunks_without_shuffle(cls, op, splits, out_shape, left, right):
+        out_chunks = []
```

Comment:
Please assign splits[axis][left_or_right][splits_or_back_mappings] explicitly to variables before referencing them inside the loop. splits[a][b][c][d] makes it hard for people to understand.

Reply:
ditto